### DISCUSSION TITLE
Option to to minimize splitting the power on phases the deferrable loads to a minimum

### DISCUSSION THREAD

**Discussion by toggm** [davidusb-geek/emhass#294] [Q&A]
I've configured two deferrable loads together with solcast and mlforecaster to calculate the day-ahead optimizations. One deferrable load is non semi-continuous variable with 4kw and 6 hours to operate. The second one is semi-continuous with 4kw and 4 hours to operate.
 Running emhass I recognized that the on-phased especially of the non semi-continuous deferrable load is splitted into multiple time segments even if there would be enough energy to continuously operate the deferrable in a single time window and attach the second one after it.
 As far as I know it is suggested to power-on and power-off a heatpump as less as possible as it might have an impact on the lifetime of the heatpump.
 
 The second thing I recognized that parts of the splitted on time-windows are pushed to the end of the solar power production phase taking the risk that the during the end of the day the real values to diverge more often from the forecasted values.
 
 The following picture shows both behaviours:
 
![2024-05-28_17-02](https://github.com/davidusb-geek/emhass/assets/3910363/11e8eea5-2985-478d-abb0-45bb8e21fb0d)

Is there a possiblity to control the number of times a deferrable load and or should be turned on and off again? Do you have other suggestions to optimize the forecasted values?

Values integrated in HA:
![2024-05-28_17-18](https://github.com/davidusb-geek/emhass/assets/3910363/4d648233-f23d-40a2-b4af-f7b343bae83d)

**Comment by davidusb-geek** [comment]
Did you tried using the `set_def_constant` parameter. Setting it to `True` should lead to just one startup of that deferrable load?
Yes for heat pumps it is advised to not turn them on and off very often.

**Reply by toggm** [reply]
Thanks for your answer and your great work. Going to try it out.

**Reply by toggm** [reply]
now it looks better, but still wondering why the on-time are very close and even overlapping start and end of solar power production when in-between there would be much more reserve:
![image](https://github.com/davidusb-geek/emhass/assets/3910363/fa9fb66d-ec86-4a2e-be1d-abcb8a56713c)

at least the second defereable on time could be closer to the end of the first one.

**Reply by davidusb-geek** [reply]
Well it depends on the input data that your are using. That result is the optimal convergence of the algorithm based on your input data. For example, what is your input data for production price and power consumption cost?
Also just to be sure always check that the algorithm actually converge: check that status is `optimal`

**Reply by toggm** [reply]
![image](https://github.com/davidusb-geek/emhass/assets/3910363/0d41811f-c81f-443d-a048-c58e2e074c3f)
![image](https://github.com/davidusb-geek/emhass/assets/3910363/9d851aad-64e9-48b8-9511-243c8acad100)

Configuration of prices and consumption costs are:
``` 
  weather_forecast_method: 'solcast'
  load_forecast_method: 'mlforecaster'
  load_cost_forecast_method: 'hp_hc_periods'
  list_hp_periods: 
    - period_hp_1:
      - start: '07:00'
      - end: '21:00'
  load_cost_hp: 0.32
  load_cost_hc: 0.21
  prod_price_forecast_method: 'constant'
  prod_sell_price: 0.145

```

**Reply by toggm** [reply]
cost optimization would be the same if the second deferrable would be powered on earlier the day. At the end of the day or close to the threshold of the forecasted power production the uncertainty is just higher that the real value will not match the forecasted one.

**Comment by davidusb-geek** [comment]
Your prod price seems pretty high for me. So that would explain why is trying to maximize profit by selling PV to the grid.
If your are sure about your input data and the optimization status is optimal then we typically have confidence on these results.
But you can still check all this results yourself if you want to be absolutely sure, that's why that big table is there. Take the results table and put it in an Excel spreadsheet and verify yourself what would have been that total cost if the system would do what it was intuitive for you to do: in the case of your reasoning the intuitive thing to do was to schedule those deferrable load when PV production was the highest. You can manually check that on Excel and actually find that if your are looking to maximize profit, then the results from the algorithm are correct.

**Reply by toggm** [reply]
Well, I'm quite confident that the result would be the same moving the second deferrable load to an earlier point in time as the price of selling would be still the same. As long as the forecast will match the real values of the production, the system itself is correct. But in case the forecast will not be the same the deferrable load would need to consume power from the grid at the an even higher rate which would then lead to non-optimal result anymore.
I'm thinking about adding a kind of uncertainty factor to the calculation of the prices the more self-production vs consumption approaches 0. 
This should lead the system to power on deferrables earlier in my case as it doesn't make a difference to power on them earlier but reducing the risk that at the end of the day the production might not be as good as forecasted.

**Reply by toggm** [reply]
I tried it out and running the second deferrable in a row resulted in the expected same result.

**Reply by toggm** [reply]
or consider kind of risk factor when sorting equal results.
